---
title: "Class 5: Genomic Interval Analysis part 2"
author: "Jay Hesselberth"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Class-5}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "img/class-2-",
  fig.align = "center"
)
```


# Genomic Interval Data pt.2

  - Review exercises 3
  - Generating randomized intervals with `bed_shuffle()` and `bed_random()`
  - Metagene profiles with multiple samples
  - Using functions to simplify analysis workflows
  - Using lists to organize many R objects
  

## Generating random intervals using `bed_shuffle()` and `bed_random()`
  In the weekend assignment you were asked to generate a metagene profile of RNAPolII ChIP-Seq density surrounding the transcription start site (TSS) of human genes. The plot obtained suggested that RNAPolII seems to be enriched near the TSS. One way to strengthen this argument would be to determine what this profile would look like if we randomly placed RNAPolII binding sites around the genome. 
  
  `valr` provides two functions for randomzing intervals. `bed_random()` which generates intervals of a fixed length throughout the genome and `bed_shuffle()` which shuffles the location of a set of predefined intervals, but maintains each intervals length.  
  
```{r}
library(valr)
library(tidyverse)
library(eda)

genome_file <- system.file("extdata", "hg19_genome.txt.gz", package = "eda")
genome <- read_genome(genome_file)


bed <- read_bed("https://de.cyverse.org/anon-files/iplant/home/kriemo/data/class/polra2_hela_1e6.bg.gz",
                n_fields = 4)
bg <- select(bed, everything(), value = name)
bg <- mutate(bg, 
              value = as.numeric(value))

random_ivls <- bed_random(genome)

shuffled_ivls <- bed_shuffle(bg, genome)
```


## Exercises

1) Generate 100,000 random intervals of 100 bp each. Set the random number generator seed to `42` Use the `genome` R object above. Which chromosome has the most intervals?

2) Assign the random intervals produced in question 1 to a new object called `random_ivls`. Shuffle these intervals using `bed_shuffle()`, and supply a seed set to `42`. How many intervals are now on `chr1`? Shuffle the random intervals (`random_ivls`) again, but set the `within` argument to `TRUE`. What effect does this have on the number of intervals per chromosome?

3) Use `bed_shuffle()` to generate a set of shuffled intervals for the RNA-PolII ChIP-Seq bedgraphs. Again set the seed argument to `42`. Assign the output to an object named `shuffled_coverage`. Calculate the total number of basepairs covered in the shuffled and the original bed tibbles. 

## How to generate a metagene plot with multiple samples?

In the previous exercise we generated a shuffled bedgraph (`shuffled_ivls`). Using this as our `bedgraph` file (object `bg`), we can recompute our metagene to examine the distribution of randomly positioned RNA-PolII. But how would we plot the shuffled and the original ChIP-seq data on the same graph?

Shown below is an example metagene profile computed for 1 bedgraph. Each `.win_id` corresponds to one `bin` or `window` region surrounding the `TSS`. The first `.win_id` represents the mean coverage from the region -1000 to -990 for all TSS in the genome. The second `.win_id` represents regions -990 to -980 and so on. 

How would we modify our tibble to be able to plot multiple ChIP-Seq experiments?

```r
#> # A tibble: 201 x 2
#>    .win_id total_coverage
#>      <int>          <dbl>
#>  1       1       23.88922
#>  2       2       26.19525
#>  3       3       23.18755
#>  4       4       22.34402
#>  5       5       26.30945
#>  6       6       28.71125
#>  7       7       28.66961
#>  8       8       23.27538
#>  9       9       35.06543
#> 10      10       34.67965
#> # ... with 191 more rows
```

If we add another column that indicates the experiment, we can `facet`, or apply an `aesthetic` based on the experiment. 

```r
#> # A tibble: 402 x 3
#>    .win_id total_coverage experiment
#>      <int>          <dbl>      <dbl>
#>  1       1       23.88922          1
#>  2       1       23.88922          2
#>  3       2       26.19525          1
#>  4       2       26.19525          2
#>  5       3       23.18755          1
#>  6       3       23.18755          2
#>  7       4       22.34402          1
#>  8       4       22.34402          2
#>  9       5       26.30945          1
#> 10       5       26.30945          2
#> # ... with 392 more rows
```

## bind_rows(), binds rows of multiple tibbles together

The `dplyr` function `bind_rows()` combines multiple tibbles together in a row-wise fashion. `bind_cols` binds tibbles together column-wise.

```{r, bind_rows}

tibble_1 <- data_frame(id = seq(1, 100),
  measurement_1 = rnorm(100))
tibble_1

tibble_2 <- data_frame(id = seq(1, 100),
  measurement_1 = rnorm(100))
tibble_2

bind_rows(tibble_1, tibble_2)

bind_rows(tibble_1, tibble_2, .id = "tibble_id")
```

## Generating a metagene with multiple samples

First let's read all of our bed files into `R`
```{r}

genes_file <- system.file("extdata", "hg19_genes.bed.gz", package = "eda")
genes <- read_bed(genes_file, n_fields = 6)

genome_file <- system.file("extdata", "hg19_genome.txt.gz", package = "eda")
genome <- read_genome(genome_file)

bed <- read_bed("https://de.cyverse.org/anon-files/iplant/home/kriemo/data/class/polra2_hela_1e6.bg.gz",
                n_fields = 4)
bg <- select(bed, everything(), value = name)
bg <- mutate(bg, 
              value = as.numeric(value))
```

Then let's generate our set of windows surrounding each TSS ("+" strand only)
```{r generate windows}
tss_windows <- genes %>% 
  filter(strand == "+") %>% 
  mutate(end = start + 1) %>% 
  bed_slop(genome, both = 1000) %>% 
  bed_makewindows(genome, win_size = 10)
```

Next, let's compute coverage for the rnapolII bedgraph. 

```{r compute_coverage}
rnapolII <- bed_map(tss_windows, 
          bg,
          total_reads = sum(value)) %>% 
  group_by(.win_id) %>%
  summarize(total_coverage = mean(total_reads, na.rm = TRUE)) 
```

Lastly, let's compute coverage for a shuffled bedgraph. 


```{r}

shuffled_coverage <- bed_shuffle(bg, genome, seed = 42)

shuffled <- bed_map(tss_windows, 
          shuffled_coverage,
          total_reads = sum(value)) %>% 
  group_by(.win_id) %>%
  summarize(total_coverage = mean(total_reads, na.rm = TRUE)) 

rnapolII

shuffled
```

Now we can bind these results together and we are ready to plot:

```{r}
res <- bind_rows(rnapolII,
          shuffled,
          .id = "expt")
res
```

Add finally plot the results
```{r}

ggplot(res, aes(.win_id, total_coverage)) +
  geom_point(aes(colour = expt))

```

## Writing a function to produce a metagene

At this point we have generated profiles for the RNAPolII chip and a shuffled control. As you may have noticed, a lot of the code that we wrote was duplicated.

Here's all of the code used to compute the metagene profile. Imagine that we wanted to now generate additional metagene profiles. We could copy and paste the code below and change it to make a new plot. What parts of the code would we need to change in order to make a metagene for a new bedGraph file? 

```
output <- genes %>% 
  filter(strand == "+") %>% 
  mutate(end = start + 1) %>% 
  bed_slop(genome, both = 1000) %>% 
  bed_makewindows(genome, win_size = 10) %>% 
  bed_map(., 
          bg,
          total_reads = sum(value)) %>% 
  group_by(.win_id) %>%
  summarize(total_coverage = mean(total_reads, na.rm = TRUE)) 
```

As you can see all that we need to change is the input `bg` object. 

## Writing functions reduces repeated code

A common rule in programming is `DRY` or **D**on't **R**epeat **Y**ourself. If you need to do something more than twice, then write a function. This reduces the amount of code you have to type, and therefore reduces typos. Additionally, if you need to change something, now you dont have to change it everywhere you copied and pasted the code, but rather just in the function.

## Introduction to functions in R

```{r}
hello_world <- function(){
  print("hello world")
}

hello_world()
```

```{r add_fxn}
add_one_to_one <- function(){
  1 + 1
}

add_one_to_one()

add_one <- function(number_to_add){
  1 + number_to_add
}

add_one(42)

int_to_add <- 42
add_one(int_to_add)

```

## Functions can have default arguments

```{r}
add_any <- function(first = 1, second = 2){
  res <- first + second
  res
}

add_any()
add_any(4)
add_any(2, 2)
```


## Computing the metagene profile using a function

When writing a function we need to identify all of the objects that need to be passed to the function. A function should not depend on objects in in the environment (i.e. current session) to run.

For example in the following code, `genes`, `genome`, and `bg` all need to be in the environment for this code to run. Therefore we should make these arguments that are passed to the function. 
```{r, eval = F}
output <- genes %>% 
  filter(strand == "+") %>% 
  mutate(end = start + 1) %>% 
  bed_slop(genome, both = 1000) %>% 
  bed_makewindows(genome, win_size = 10) %>% 
  bed_map(., 
          bg,
          total_reads = sum(value)) %>% 
  group_by(.win_id) %>%
  summarize(total_coverage = mean(total_reads, na.rm = TRUE)) 

```

Here, we've designated the `genes`, `genome` and `bg` objects as function arguments and renamed them to make it clearer what the identity of each argument is.  

```{r}
compute_metagene <- function(chip_seq_bedgraph,
                                genes_bed,
                              genome_bed){
  output <- genes_bed %>% 
    filter(strand == "+") %>% 
    mutate(end = start + 1) %>% 
    bed_slop(genome_bed, both = 1000) %>% 
    bed_makewindows(genome_bed, win_size = 10) %>% 
    bed_map(., 
            chip_seq_bedgraph,
            total_reads = sum(value)) %>% 
    group_by(.win_id) %>%
    summarize(total_coverage = mean(total_reads, na.rm = TRUE)) 
  output
}

compute_metagene(bg, genes, genome)
```

Now we can apply the function to any arbitrary bedgraph:

```{r eval = F}
compute_metagene(shuffled_coverage, genes, genome)

compute_metagene(..., genes, genome)
```

## Using lists to organize many objects

Imagine we perform our ChIP-Seq experiment with many replicates. We can use the `read_bed()` to read all of these files into `R`, but now we have a whole bunch of `R` objects in our environment.

```
bg_chipseq_rep1 <- read_bed("rep1.bg.gz")
bg_chipseq_rep2 <- read_bed("rep2.bg.gz")
bg_chipseq_rep3 <- read_bed("rep3.bg.gz")
bg_chipseq_rep4 <- read_bed("rep4.bg.gz")
bg_chipseq_rep5 <- read_bed("rep5.bg.gz")
bg_chipseq_rep6 <- read_bed("rep6.bg.gz")
```

You can imagine that we could end up generating many intermediate files, and again repeating ourselves, even though we are using a function. 

```r
metagene_chipseq_rep1 <- compute_metagene(bg_chipseq_rep1, genome, genes)
metagene_chipseq_rep2 <- compute_metagene(bg_chipseq_rep2, genome, genes)
metagene_chipseq_rep3 <- compute_metagene(bg_chipseq_rep3, genome, genes)
metagene_chipseq_rep4 <- compute_metagene(bg_chipseq_rep4, genome, genes)
metagene_chipseq_rep5 <- compute_metagene(bg_chipseq_rep5, genome, genes)
metagene_chipseq_rep6 <- compute_metagene(bg_chipseq_rep6, genome, genes)
```

## Lists make it easy to store and organize many objects

```r
chipseq_bgs <- list(bg_chipseq_rep1,
                    bg_chipseq_rep2,
                    bg_chipseq_rep3,
                    bg_chipseq_rep4,
                    bg_chipseq_rep5,
                    bg_chipseq_rep6)

chipseq_bgs
```

You can access an individual element by index

```r
chipseq_bgs[[1]] # first object
chipseq_bgs[[2]] # second object
```

Or you can name each element, and access by name using the `$` symbol

```r
chipseq_bgs <- list(rep1 = bg_chipseq_rep1,
                    rep2 = bg_chipseq_rep2,
                    rep3 = bg_chipseq_rep3,
                    rep4 = bg_chipseq_rep4,
                    rep5 = bg_chipseq_rep5,
                    rep6 = bg_chipseq_rep6)
                 
# access first tibble                    
chipseq_bgs$rep1 

# print chrom column from first tibble
chipseq_bgs$rep1$chrom

select(chipseq_bgs$rep1, chrom)
```

Now we can refer to a single list object:

```
metagenes_chipseq = list()
metagenes_chipseq$rep1 <- compute_metagene(chipseq_bgs$rep1, genome, genes)
metagenes_chipseq$rep2 <- compute_metagene(chipseq_bgs$rep2, genome, genes)
...
metagenes_chipseq$rep6 <- compute_metagene(chipseq_bgs$rep6, genome, genes)
```

## The map function 

The `map()` function takes a list or vector and applys a function to each element of the list or vector.

```
map(a_vector, a_function)

my_numbers <- c(1, 2, 3)

add_one

map(my_numbers, add_one)
map(my_numbers, ~add_one(.x))
```

This is extremely useful, as it allows us to apply a function to any number of R objects. In our toy example, to compute a metagene profile for the six chip-seq experiments, we now only need one line.

```r
metagenes_chipses <- map(chipseq_bgs, ~compute_metagene(.x, genome, genes))
```


Here's a live example using shuffled bedgraphs for simplicity:

```{r map_example}

# make smaller bg 
small_bg <- sample_n(bg, 1e5)

bg1 <- bed_shuffle(small_bg, genome)
bg2 <- bed_shuffle(small_bg, genome)
bg3 <- bed_shuffle(small_bg, genome)

bgs <- list(bg1, 
            bg2,
            bg3)
            
res <- map(bgs, ~compute_metagene(.x, genes, genome))
res
```


## Combining many tibbles

`bind_rows()` can bind any number of tibbles if the tibbles are supplied as a `list()`. This is very useful to combining many dataframes that are in a list into a format for plotting. 

```{r}
all_res <- bind_rows(res, .id = "expt")
all_res
```

```{r plot_random}
ggplot(all_res, aes(.win_id, total_coverage)) +
  geom_point(aes(colour = expt))

```

## Exercises

The exercises for today's class are presented [here](https://rnabioco.github.io/eda/articles/exercises-4.html), Here is a [link](https://raw.githubusercontent.com/rnabioco/eda/master/vignettes/exercises-4.Rmd) to the raw text of these exercises

## Quiz

Create an RMarkdown document and answer the questions from the Exercises section. Write your answers in the text, and print the final tibbles that your code produced that gave you the answer. Submit your final document as "Quiz 4" by **Tuesday at 10 PM**.

**Your submitted document must knit to HTML without errors**. I.e., when you click the "Knit" button, the document should build and display and HTML page. 
