<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Class 5: Genomic Interval Analysis part 2 • eda</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Google analytics --><script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-54095998-4', 'auto');
  ga('send', 'pageview');

</script>
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">eda</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Class Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/class-1.html">Class 1: Overview and Data Tidying</a>
    </li>
    <li>
      <a href="../articles/class-2.html">Class 2: Writing Reproducible Reports</a>
    </li>
    <li>
      <a href="../articles/class-3.html">Class 3: Common Analysis Challenges</a>
    </li>
    <li>
      <a href="../articles/class-4.html">Class 4: Genomic Interval Analysis part 1</a>
    </li>
    <li>
      <a href="../articles/class-5.html">Class 5: Genomic Interval Analysis part 2</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/exercises-1.html">Exercises 1</a>
    </li>
    <li>
      <a href="../articles/exercises-2.html">Exercises 2</a>
    </li>
    <li>
      <a href="../articles/exercises-3.html">Exercises 3</a>
    </li>
    <li>
      <a href="../articles/exercises-4.html">Exercises 4</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignment-1.html">Weekend Assignment 1</a>
    </li>
    <li>
      <a href="../articles/assignment-2.html">Weekend Assignment 2</a>
    </li>
    <li class="divider">
    <li>
      <a href="../articles/assignments.html">Assignment Submission and Grading Rubric</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Class 5: Genomic Interval Analysis part 2</h1>
                        <h4 class="author">Jay Hesselberth</h4>
            
            <h4 class="date">2017-12-11</h4>
          </div>

    
    
<div class="contents">
<div id="genomic-interval-data-pt-2" class="section level1">
<h1 class="hasAnchor">
<a href="#genomic-interval-data-pt-2" class="anchor"></a>Genomic Interval Data pt.2</h1>
<ul>
<li>Review exercises 3</li>
<li>Generating randomized intervals with <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle()</a></code> and <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_random">bed_random()</a></code>
</li>
<li>Metagene profiles with multiple samples</li>
<li>Using functions to simplify analysis workflows</li>
<li>Using lists to organize many R objects</li>
</ul>
<div id="generating-random-intervals-using-bed_shuffle-and-bed_random" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-random-intervals-using-bed_shuffle-and-bed_random" class="anchor"></a>Generating random intervals using <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle()</a></code> and <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_random">bed_random()</a></code>
</h2>
<p>In the weekend assignment you were asked to generate a metagene profile of RNAPolII ChIP-Seq density surrounding the transcription start site (TSS) of human genes. The plot obtained suggested that RNAPolII seems to be enriched near the TSS. One way to strengthen this argument would be to determine what this profile would look like if we randomly placed RNAPolII binding sites around the genome.</p>
<p><code>valr</code> provides two functions for randomzing intervals. <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_random">bed_random()</a></code> which generates intervals of a fixed length throughout the genome and <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle()</a></code> which shuffles the location of a set of predefined intervals, but maintains each intervals length.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
<span class="kw">library</span>(tidyverse)
<span class="co">#&gt; ── Attaching packages ────────────────────────────────── tidyverse 1.2.1 ──</span>
<span class="co">#&gt; ✔ ggplot2 2.2.1          ✔ purrr   0.2.4     </span>
<span class="co">#&gt; ✔ tibble  1.3.4          ✔ dplyr   0.7.4.9000</span>
<span class="co">#&gt; ✔ tidyr   0.7.2          ✔ stringr 1.2.0     </span>
<span class="co">#&gt; ✔ readr   1.1.1          ✔ forcats 0.2.0</span>
<span class="co">#&gt; ── Conflicts ───────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">#&gt; ✖ dplyr::filter() masks stats::filter()</span>
<span class="co">#&gt; ✖ dplyr::lag()    masks stats::lag()</span>
<span class="kw">library</span>(eda)

genome_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"hg19_genome.txt.gz"</span>, <span class="dt">package =</span> <span class="st">"eda"</span>)
genome &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_genome">read_genome</a></span>(genome_file)


bed &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(<span class="st">"https://de.cyverse.org/anon-files/iplant/home/kriemo/data/class/polra2_hela_1e6.bg.gz"</span>,
                <span class="dt">n_fields =</span> <span class="dv">4</span>)
bg &lt;-<span class="st"> </span><span class="kw">select</span>(bed, <span class="kw">everything</span>(), <span class="dt">value =</span> name)
bg &lt;-<span class="st"> </span><span class="kw">mutate</span>(bg, 
              <span class="dt">value =</span> <span class="kw">as.numeric</span>(value))

random_ivls &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_random">bed_random</a></span>(genome)

shuffled_ivls &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle</a></span>(bg, genome)</code></pre></div>
</div>
<div id="exercises" class="section level2">
<h2 class="hasAnchor">
<a href="#exercises" class="anchor"></a>Exercises</h2>
<ol style="list-style-type: decimal">
<li><p>Generate 100,000 random intervals of 100 bp each. Set the random number generator seed to <code>42</code> Use the <code>genome</code> R object above. Which chromosome has the most intervals?</p></li>
<li><p>Assign the random intervals produced in question 1 to a new object called <code>random_ivls</code>. Shuffle these intervals using <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle()</a></code>, and supply a seed set to <code>42</code>. How many intervals are now on <code>chr1</code>? Shuffle the random intervals (<code>random_ivls</code>) again, but set the <code>within</code> argument to <code>TRUE</code>. What effect does this have on the number of intervals per chromosome?</p></li>
<li><p>Use <code><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle()</a></code> to generate a set of shuffled intervals for the RNA-PolII ChIP-Seq bedgraphs. Again set the seed argument to <code>42</code>. Assign the output to an object named <code>shuffled_coverage</code>. Calculate the total number of basepairs covered in the shuffled and the original bed tibbles.</p></li>
</ol>
</div>
<div id="how-to-generate-a-metagene-plot-with-multiple-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-generate-a-metagene-plot-with-multiple-samples" class="anchor"></a>How to generate a metagene plot with multiple samples?</h2>
<p>In the previous exercise we generated a shuffled bedgraph (<code>shuffled_ivls</code>). Using this as our <code>bedgraph</code> file (object <code>bg</code>), we can recompute our metagene to examine the distribution of randomly positioned RNA-PolII. But how would we plot the shuffled and the original ChIP-seq data on the same graph?</p>
<p>Shown below is an example metagene profile computed for 1 bedgraph. Each <code>.win_id</code> corresponds to one <code>bin</code> or <code>window</code> region surrounding the <code>TSS</code>. The first <code>.win_id</code> represents the mean coverage from the region -1000 to -990 for all TSS in the genome. The second <code>.win_id</code> represents regions -990 to -980 and so on.</p>
<p>How would we modify our tibble to be able to plot multiple ChIP-Seq experiments?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#&gt; # A tibble: 201 x 2</span>
<span class="co">#&gt;    .win_id total_coverage</span>
<span class="co">#&gt;      &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1       23.88922</span>
<span class="co">#&gt;  2       2       26.19525</span>
<span class="co">#&gt;  3       3       23.18755</span>
<span class="co">#&gt;  4       4       22.34402</span>
<span class="co">#&gt;  5       5       26.30945</span>
<span class="co">#&gt;  6       6       28.71125</span>
<span class="co">#&gt;  7       7       28.66961</span>
<span class="co">#&gt;  8       8       23.27538</span>
<span class="co">#&gt;  9       9       35.06543</span>
<span class="co">#&gt; 10      10       34.67965</span>
<span class="co">#&gt; # ... with 191 more rows</span></code></pre></div>
<p>If we add another column that indicates the experiment, we can <code>facet</code>, or apply an <code>aesthetic</code> based on the experiment.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#&gt; # A tibble: 402 x 3</span>
<span class="co">#&gt;    .win_id total_coverage experiment</span>
<span class="co">#&gt;      &lt;int&gt;          &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1       23.88922          1</span>
<span class="co">#&gt;  2       1       23.88922          2</span>
<span class="co">#&gt;  3       2       26.19525          1</span>
<span class="co">#&gt;  4       2       26.19525          2</span>
<span class="co">#&gt;  5       3       23.18755          1</span>
<span class="co">#&gt;  6       3       23.18755          2</span>
<span class="co">#&gt;  7       4       22.34402          1</span>
<span class="co">#&gt;  8       4       22.34402          2</span>
<span class="co">#&gt;  9       5       26.30945          1</span>
<span class="co">#&gt; 10       5       26.30945          2</span>
<span class="co">#&gt; # ... with 392 more rows</span></code></pre></div>
</div>
<div id="bind_rows-binds-rows-of-multiple-tibbles-together" class="section level2">
<h2 class="hasAnchor">
<a href="#bind_rows-binds-rows-of-multiple-tibbles-together" class="anchor"></a>bind_rows(), binds rows of multiple tibbles together</h2>
<p>The <code>dplyr</code> function <code>bind_rows()</code> combines multiple tibbles together in a row-wise fashion. <code>bind_cols</code> binds tibbles together column-wise.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
tibble_1 &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">id =</span> <span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">100</span>),
  <span class="dt">measurement_1 =</span> <span class="kw">rnorm</span>(<span class="dv">100</span>))
tibble_1
<span class="co">#&gt; # A tibble: 100 x 2</span>
<span class="co">#&gt;       id measurement_1</span>
<span class="co">#&gt;    &lt;int&gt;         &lt;dbl&gt;</span>
<span class="co">#&gt;  1     1    0.66679462</span>
<span class="co">#&gt;  2     2    0.24049896</span>
<span class="co">#&gt;  3     3    1.55512874</span>
<span class="co">#&gt;  4     4   -0.75274870</span>
<span class="co">#&gt;  5     5    0.80749442</span>
<span class="co">#&gt;  6     6    0.25470467</span>
<span class="co">#&gt;  7     7    0.02687301</span>
<span class="co">#&gt;  8     8    1.53449670</span>
<span class="co">#&gt;  9     9    1.28550345</span>
<span class="co">#&gt; 10    10   -0.61429068</span>
<span class="co">#&gt; # ... with 90 more rows</span>

tibble_2 &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">id =</span> <span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">100</span>),
  <span class="dt">measurement_1 =</span> <span class="kw">rnorm</span>(<span class="dv">100</span>))
tibble_2
<span class="co">#&gt; # A tibble: 100 x 2</span>
<span class="co">#&gt;       id measurement_1</span>
<span class="co">#&gt;    &lt;int&gt;         &lt;dbl&gt;</span>
<span class="co">#&gt;  1     1     1.4841634</span>
<span class="co">#&gt;  2     2     0.5934025</span>
<span class="co">#&gt;  3     3     0.1297867</span>
<span class="co">#&gt;  4     4    -0.4987656</span>
<span class="co">#&gt;  5     5     0.2150065</span>
<span class="co">#&gt;  6     6     0.4736234</span>
<span class="co">#&gt;  7     7    -1.6283166</span>
<span class="co">#&gt;  8     8    -1.4868916</span>
<span class="co">#&gt;  9     9    -0.8753699</span>
<span class="co">#&gt; 10    10     1.4377516</span>
<span class="co">#&gt; # ... with 90 more rows</span>

<span class="kw">bind_rows</span>(tibble_1, tibble_2)
<span class="co">#&gt; # A tibble: 200 x 2</span>
<span class="co">#&gt;       id measurement_1</span>
<span class="co">#&gt;    &lt;int&gt;         &lt;dbl&gt;</span>
<span class="co">#&gt;  1     1    0.66679462</span>
<span class="co">#&gt;  2     2    0.24049896</span>
<span class="co">#&gt;  3     3    1.55512874</span>
<span class="co">#&gt;  4     4   -0.75274870</span>
<span class="co">#&gt;  5     5    0.80749442</span>
<span class="co">#&gt;  6     6    0.25470467</span>
<span class="co">#&gt;  7     7    0.02687301</span>
<span class="co">#&gt;  8     8    1.53449670</span>
<span class="co">#&gt;  9     9    1.28550345</span>
<span class="co">#&gt; 10    10   -0.61429068</span>
<span class="co">#&gt; # ... with 190 more rows</span>

<span class="kw">bind_rows</span>(tibble_1, tibble_2, <span class="dt">.id =</span> <span class="st">"tibble_id"</span>)
<span class="co">#&gt; # A tibble: 200 x 3</span>
<span class="co">#&gt;    tibble_id    id measurement_1</span>
<span class="co">#&gt;        &lt;chr&gt; &lt;int&gt;         &lt;dbl&gt;</span>
<span class="co">#&gt;  1         1     1    0.66679462</span>
<span class="co">#&gt;  2         1     2    0.24049896</span>
<span class="co">#&gt;  3         1     3    1.55512874</span>
<span class="co">#&gt;  4         1     4   -0.75274870</span>
<span class="co">#&gt;  5         1     5    0.80749442</span>
<span class="co">#&gt;  6         1     6    0.25470467</span>
<span class="co">#&gt;  7         1     7    0.02687301</span>
<span class="co">#&gt;  8         1     8    1.53449670</span>
<span class="co">#&gt;  9         1     9    1.28550345</span>
<span class="co">#&gt; 10         1    10   -0.61429068</span>
<span class="co">#&gt; # ... with 190 more rows</span></code></pre></div>
</div>
<div id="generating-a-metagene-with-multiple-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-a-metagene-with-multiple-samples" class="anchor"></a>Generating a metagene with multiple samples</h2>
<p>First let’s read all of our bed files into <code>R</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
genes_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"hg19_genes.bed.gz"</span>, <span class="dt">package =</span> <span class="st">"eda"</span>)
genes &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(genes_file, <span class="dt">n_fields =</span> <span class="dv">6</span>)

genome_file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"hg19_genome.txt.gz"</span>, <span class="dt">package =</span> <span class="st">"eda"</span>)
genome &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_genome">read_genome</a></span>(genome_file)

bed &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed</a></span>(<span class="st">"https://de.cyverse.org/anon-files/iplant/home/kriemo/data/class/polra2_hela_1e6.bg.gz"</span>,
                <span class="dt">n_fields =</span> <span class="dv">4</span>)
bg &lt;-<span class="st"> </span><span class="kw">select</span>(bed, <span class="kw">everything</span>(), <span class="dt">value =</span> name)
bg &lt;-<span class="st"> </span><span class="kw">mutate</span>(bg, 
              <span class="dt">value =</span> <span class="kw">as.numeric</span>(value))</code></pre></div>
<p>Then let’s generate our set of windows surrounding each TSS (“+” strand only)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tss_windows &lt;-<span class="st"> </span>genes %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">filter</span>(strand ==<span class="st"> "+"</span>) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">end =</span> start +<span class="st"> </span><span class="dv">1</span>) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_slop">bed_slop</a></span>(genome, <span class="dt">both =</span> <span class="dv">1000</span>) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_makewindows">bed_makewindows</a></span>(genome, <span class="dt">win_size =</span> <span class="dv">10</span>)</code></pre></div>
<p>Next, let’s compute coverage for the rnapolII bedgraph.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rnapolII &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map</a></span>(tss_windows, 
          bg,
          <span class="dt">total_reads =</span> <span class="kw">sum</span>(value)) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(.win_id) %&gt;%
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">total_coverage =</span> <span class="kw">mean</span>(total_reads, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) </code></pre></div>
<p>Lastly, let’s compute coverage for a shuffled bedgraph.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
shuffled_coverage &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle</a></span>(bg, genome, <span class="dt">seed =</span> <span class="dv">42</span>)

shuffled &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map</a></span>(tss_windows, 
          shuffled_coverage,
          <span class="dt">total_reads =</span> <span class="kw">sum</span>(value)) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(.win_id) %&gt;%
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">total_coverage =</span> <span class="kw">mean</span>(total_reads, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) 

rnapolII
<span class="co">#&gt; # A tibble: 201 x 2</span>
<span class="co">#&gt;    .win_id total_coverage</span>
<span class="co">#&gt;      &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1       23.88922</span>
<span class="co">#&gt;  2       2       26.19525</span>
<span class="co">#&gt;  3       3       23.18755</span>
<span class="co">#&gt;  4       4       22.34402</span>
<span class="co">#&gt;  5       5       26.30945</span>
<span class="co">#&gt;  6       6       28.71125</span>
<span class="co">#&gt;  7       7       28.66961</span>
<span class="co">#&gt;  8       8       23.27538</span>
<span class="co">#&gt;  9       9       35.06543</span>
<span class="co">#&gt; 10      10       34.67965</span>
<span class="co">#&gt; # ... with 191 more rows</span>

shuffled
<span class="co">#&gt; # A tibble: 201 x 2</span>
<span class="co">#&gt;    .win_id total_coverage</span>
<span class="co">#&gt;      &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1       2.998299</span>
<span class="co">#&gt;  2       2       4.265660</span>
<span class="co">#&gt;  3       3       2.352762</span>
<span class="co">#&gt;  4       4       3.682922</span>
<span class="co">#&gt;  5       5       2.937430</span>
<span class="co">#&gt;  6       6       3.082288</span>
<span class="co">#&gt;  7       7       1.839423</span>
<span class="co">#&gt;  8       8       2.165887</span>
<span class="co">#&gt;  9       9       3.044269</span>
<span class="co">#&gt; 10      10       2.049153</span>
<span class="co">#&gt; # ... with 191 more rows</span></code></pre></div>
<p>Now we can bind these results together and we are ready to plot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw">bind_rows</span>(rnapolII,
          shuffled,
          <span class="dt">.id =</span> <span class="st">"expt"</span>)
res
<span class="co">#&gt; # A tibble: 402 x 3</span>
<span class="co">#&gt;     expt .win_id total_coverage</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1     1       1       23.88922</span>
<span class="co">#&gt;  2     1       2       26.19525</span>
<span class="co">#&gt;  3     1       3       23.18755</span>
<span class="co">#&gt;  4     1       4       22.34402</span>
<span class="co">#&gt;  5     1       5       26.30945</span>
<span class="co">#&gt;  6     1       6       28.71125</span>
<span class="co">#&gt;  7     1       7       28.66961</span>
<span class="co">#&gt;  8     1       8       23.27538</span>
<span class="co">#&gt;  9     1       9       35.06543</span>
<span class="co">#&gt; 10     1      10       34.67965</span>
<span class="co">#&gt; # ... with 392 more rows</span></code></pre></div>
<p>Add finally plot the results</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="kw">ggplot</span>(res, <span class="kw">aes</span>(.win_id, total_coverage)) +
<span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">colour =</span> expt))</code></pre></div>
<p><img src="img/class-2-unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div id="writing-a-function-to-produce-a-metagene" class="section level2">
<h2 class="hasAnchor">
<a href="#writing-a-function-to-produce-a-metagene" class="anchor"></a>Writing a function to produce a metagene</h2>
<p>At this point we have generated profiles for the RNAPolII chip and a shuffled control. As you may have noticed, a lot of the code that we wrote was duplicated.</p>
<p>Here’s all of the code used to compute the metagene profile. Imagine that we wanted to now generate additional metagene profiles. We could copy and paste the code below and change it to make a new plot. What parts of the code would we need to change in order to make a metagene for a new bedGraph file?</p>
<pre><code>output &lt;- genes %&gt;% 
  filter(strand == "+") %&gt;% 
  mutate(end = start + 1) %&gt;% 
  bed_slop(genome, both = 1000) %&gt;% 
  bed_makewindows(genome, win_size = 10) %&gt;% 
  bed_map(., 
          bg,
          total_reads = sum(value)) %&gt;% 
  group_by(.win_id) %&gt;%
  summarize(total_coverage = mean(total_reads, na.rm = TRUE)) </code></pre>
<p>As you can see all that we need to change is the input <code>bg</code> object.</p>
</div>
<div id="writing-functions-reduces-repeated-code" class="section level2">
<h2 class="hasAnchor">
<a href="#writing-functions-reduces-repeated-code" class="anchor"></a>Writing functions reduces repeated code</h2>
<p>A common rule in programming is <code>DRY</code> or <strong>D</strong>on’t <strong>R</strong>epeat <strong>Y</strong>ourself. If you need to do something more than twice, then write a function. This reduces the amount of code you have to type, and therefore reduces typos. Additionally, if you need to change something, now you dont have to change it everywhere you copied and pasted the code, but rather just in the function.</p>
</div>
<div id="introduction-to-functions-in-r" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-functions-in-r" class="anchor"></a>Introduction to functions in R</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hello_world &lt;-<span class="st"> </span>function(){
  <span class="kw">print</span>(<span class="st">"hello world"</span>)
}

<span class="kw">hello_world</span>()
<span class="co">#&gt; [1] "hello world"</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">add_one_to_one &lt;-<span class="st"> </span>function(){
  <span class="dv">1</span> +<span class="st"> </span><span class="dv">1</span>
}

<span class="kw">add_one_to_one</span>()
<span class="co">#&gt; [1] 2</span>

add_one &lt;-<span class="st"> </span>function(number_to_add){
  <span class="dv">1</span> +<span class="st"> </span>number_to_add
}

<span class="kw">add_one</span>(<span class="dv">42</span>)
<span class="co">#&gt; [1] 43</span>

int_to_add &lt;-<span class="st"> </span><span class="dv">42</span>
<span class="kw">add_one</span>(int_to_add)
<span class="co">#&gt; [1] 43</span></code></pre></div>
</div>
<div id="functions-can-have-default-arguments" class="section level2">
<h2 class="hasAnchor">
<a href="#functions-can-have-default-arguments" class="anchor"></a>Functions can have default arguments</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">add_any &lt;-<span class="st"> </span>function(<span class="dt">first =</span> <span class="dv">1</span>, <span class="dt">second =</span> <span class="dv">2</span>){
  res &lt;-<span class="st"> </span>first +<span class="st"> </span>second
  res
}

<span class="kw">add_any</span>()
<span class="co">#&gt; [1] 3</span>
<span class="kw">add_any</span>(<span class="dv">4</span>)
<span class="co">#&gt; [1] 6</span>
<span class="kw">add_any</span>(<span class="dv">2</span>, <span class="dv">2</span>)
<span class="co">#&gt; [1] 4</span></code></pre></div>
</div>
<div id="computing-the-metagene-profile-using-a-function" class="section level2">
<h2 class="hasAnchor">
<a href="#computing-the-metagene-profile-using-a-function" class="anchor"></a>Computing the metagene profile using a function</h2>
<p>When writing a function we need to identify all of the objects that need to be passed to the function. A function should not depend on objects in in the environment (i.e. current session) to run.</p>
<p>For example in the following code, <code>genes</code>, <code>genome</code>, and <code>bg</code> all need to be in the environment for this code to run. Therefore we should make these arguments that are passed to the function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">output &lt;-<span class="st"> </span>genes %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">filter</span>(strand ==<span class="st"> "+"</span>) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">end =</span> start +<span class="st"> </span><span class="dv">1</span>) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_slop">bed_slop</a></span>(genome, <span class="dt">both =</span> <span class="dv">1000</span>) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_makewindows">bed_makewindows</a></span>(genome, <span class="dt">win_size =</span> <span class="dv">10</span>) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map</a></span>(., 
          bg,
          <span class="dt">total_reads =</span> <span class="kw">sum</span>(value)) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(.win_id) %&gt;%
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">total_coverage =</span> <span class="kw">mean</span>(total_reads, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) </code></pre></div>
<p>Here, we’ve designated the <code>genes</code>, <code>genome</code> and <code>bg</code> objects as function arguments and renamed them to make it clearer what the identity of each argument is.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">compute_metagene &lt;-<span class="st"> </span>function(chip_seq_bedgraph,
                                genes_bed,
                              genome_bed){
  output &lt;-<span class="st"> </span>genes_bed %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">filter</span>(strand ==<span class="st"> "+"</span>) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">end =</span> start +<span class="st"> </span><span class="dv">1</span>) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_slop">bed_slop</a></span>(genome_bed, <span class="dt">both =</span> <span class="dv">1000</span>) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_makewindows">bed_makewindows</a></span>(genome_bed, <span class="dt">win_size =</span> <span class="dv">10</span>) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_map">bed_map</a></span>(., 
            chip_seq_bedgraph,
            <span class="dt">total_reads =</span> <span class="kw">sum</span>(value)) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw">group_by</span>(.win_id) %&gt;%
<span class="st">    </span><span class="kw">summarize</span>(<span class="dt">total_coverage =</span> <span class="kw">mean</span>(total_reads, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) 
  output
}

<span class="kw">compute_metagene</span>(bg, genes, genome)
<span class="co">#&gt; # A tibble: 201 x 2</span>
<span class="co">#&gt;    .win_id total_coverage</span>
<span class="co">#&gt;      &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1       23.88922</span>
<span class="co">#&gt;  2       2       26.19525</span>
<span class="co">#&gt;  3       3       23.18755</span>
<span class="co">#&gt;  4       4       22.34402</span>
<span class="co">#&gt;  5       5       26.30945</span>
<span class="co">#&gt;  6       6       28.71125</span>
<span class="co">#&gt;  7       7       28.66961</span>
<span class="co">#&gt;  8       8       23.27538</span>
<span class="co">#&gt;  9       9       35.06543</span>
<span class="co">#&gt; 10      10       34.67965</span>
<span class="co">#&gt; # ... with 191 more rows</span></code></pre></div>
<p>Now we can apply the function to any arbitrary bedgraph:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">compute_metagene</span>(shuffled_coverage, genes, genome)

<span class="kw">compute_metagene</span>(..., genes, genome)</code></pre></div>
</div>
<div id="using-lists-to-organize-many-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#using-lists-to-organize-many-objects" class="anchor"></a>Using lists to organize many objects</h2>
<p>Imagine we perform our ChIP-Seq experiment with many replicates. We can use the <code><a href="http://www.rdocumentation.org/packages/valr/topics/read_bed">read_bed()</a></code> to read all of these files into <code>R</code>, but now we have a whole bunch of <code>R</code> objects in our environment.</p>
<pre><code>bg_chipseq_rep1 &lt;- read_bed("rep1.bg.gz")
bg_chipseq_rep2 &lt;- read_bed("rep2.bg.gz")
bg_chipseq_rep3 &lt;- read_bed("rep3.bg.gz")
bg_chipseq_rep4 &lt;- read_bed("rep4.bg.gz")
bg_chipseq_rep5 &lt;- read_bed("rep5.bg.gz")
bg_chipseq_rep6 &lt;- read_bed("rep6.bg.gz")</code></pre>
<p>You can imagine that we could end up generating many intermediate files, and again repeating ourselves, even though we are using a function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metagene_chipseq_rep1 &lt;-<span class="st"> </span><span class="kw">compute_metagene</span>(bg_chipseq_rep1, genome, genes)
metagene_chipseq_rep2 &lt;-<span class="st"> </span><span class="kw">compute_metagene</span>(bg_chipseq_rep2, genome, genes)
metagene_chipseq_rep3 &lt;-<span class="st"> </span><span class="kw">compute_metagene</span>(bg_chipseq_rep3, genome, genes)
metagene_chipseq_rep4 &lt;-<span class="st"> </span><span class="kw">compute_metagene</span>(bg_chipseq_rep4, genome, genes)
metagene_chipseq_rep5 &lt;-<span class="st"> </span><span class="kw">compute_metagene</span>(bg_chipseq_rep5, genome, genes)
metagene_chipseq_rep6 &lt;-<span class="st"> </span><span class="kw">compute_metagene</span>(bg_chipseq_rep6, genome, genes)</code></pre></div>
</div>
<div id="lists-make-it-easy-to-store-and-organize-many-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#lists-make-it-easy-to-store-and-organize-many-objects" class="anchor"></a>Lists make it easy to store and organize many objects</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chipseq_bgs &lt;-<span class="st"> </span><span class="kw">list</span>(bg_chipseq_rep1,
                    bg_chipseq_rep2,
                    bg_chipseq_rep3,
                    bg_chipseq_rep4,
                    bg_chipseq_rep5,
                    bg_chipseq_rep6)

chipseq_bgs</code></pre></div>
<p>You can access an individual element by index</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chipseq_bgs[[<span class="dv">1</span>]] <span class="co"># first object</span>
chipseq_bgs[[<span class="dv">2</span>]] <span class="co"># second object</span></code></pre></div>
<p>Or you can name each element, and access by name using the <code>$</code> symbol</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chipseq_bgs &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">rep1 =</span> bg_chipseq_rep1,
                    <span class="dt">rep2 =</span> bg_chipseq_rep2,
                    <span class="dt">rep3 =</span> bg_chipseq_rep3,
                    <span class="dt">rep4 =</span> bg_chipseq_rep4,
                    <span class="dt">rep5 =</span> bg_chipseq_rep5,
                    <span class="dt">rep6 =</span> bg_chipseq_rep6)
                 
<span class="co"># access first tibble                    </span>
chipseq_bgs$rep1 

<span class="co"># print chrom column from first tibble</span>
chipseq_bgs$rep1$chrom

<span class="kw">select</span>(chipseq_bgs$rep1, chrom)</code></pre></div>
<p>Now we can refer to a single list object:</p>
<pre><code>metagenes_chipseq = list()
metagenes_chipseq$rep1 &lt;- compute_metagene(chipseq_bgs$rep1, genome, genes)
metagenes_chipseq$rep2 &lt;- compute_metagene(chipseq_bgs$rep2, genome, genes)
...
metagenes_chipseq$rep6 &lt;- compute_metagene(chipseq_bgs$rep6, genome, genes)</code></pre>
</div>
<div id="the-map-function" class="section level2">
<h2 class="hasAnchor">
<a href="#the-map-function" class="anchor"></a>The map function</h2>
<p>The <code>map()</code> function takes a list or vector and applys a function to each element of the list or vector.</p>
<pre><code>map(a_vector, a_function)

my_numbers &lt;- c(1, 2, 3)

add_one

map(my_numbers, add_one)
map(my_numbers, ~add_one(.x))</code></pre>
<p>This is extremely useful, as it allows us to apply a function to any number of R objects. In our toy example, to compute a metagene profile for the six chip-seq experiments, we now only need one line.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">metagenes_chipses &lt;-<span class="st"> </span><span class="kw">map</span>(chipseq_bgs, ~<span class="kw">compute_metagene</span>(.x, genome, genes))</code></pre></div>
<p>Here’s a live example using shuffled bedgraphs for simplicity:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="co"># make smaller bg </span>
small_bg &lt;-<span class="st"> </span><span class="kw">sample_n</span>(bg, <span class="fl">1e5</span>)

bg1 &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle</a></span>(small_bg, genome)
bg2 &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle</a></span>(small_bg, genome)
bg3 &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/valr/topics/bed_shuffle">bed_shuffle</a></span>(small_bg, genome)

bgs &lt;-<span class="st"> </span><span class="kw">list</span>(bg1, 
            bg2,
            bg3)
            
res &lt;-<span class="st"> </span><span class="kw">map</span>(bgs, ~<span class="kw">compute_metagene</span>(.x, genes, genome))
res
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; # A tibble: 201 x 2</span>
<span class="co">#&gt;    .win_id total_coverage</span>
<span class="co">#&gt;      &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1      0.8479167</span>
<span class="co">#&gt;  2       2      4.7904762</span>
<span class="co">#&gt;  3       3      0.9065217</span>
<span class="co">#&gt;  4       4      4.7795918</span>
<span class="co">#&gt;  5       5      4.1403846</span>
<span class="co">#&gt;  6       6      2.3000000</span>
<span class="co">#&gt;  7       7      6.1677419</span>
<span class="co">#&gt;  8       8      2.9953846</span>
<span class="co">#&gt;  9       9      7.5516667</span>
<span class="co">#&gt; 10      10      0.9807018</span>
<span class="co">#&gt; # ... with 191 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; # A tibble: 201 x 2</span>
<span class="co">#&gt;    .win_id total_coverage</span>
<span class="co">#&gt;      &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1       4.665517</span>
<span class="co">#&gt;  2       2       9.992857</span>
<span class="co">#&gt;  3       3       3.770492</span>
<span class="co">#&gt;  4       4       2.020000</span>
<span class="co">#&gt;  5       5       2.292727</span>
<span class="co">#&gt;  6       6       6.014286</span>
<span class="co">#&gt;  7       7       1.310909</span>
<span class="co">#&gt;  8       8       2.226667</span>
<span class="co">#&gt;  9       9       4.430435</span>
<span class="co">#&gt; 10      10       0.887500</span>
<span class="co">#&gt; # ... with 191 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[3]]</span>
<span class="co">#&gt; # A tibble: 201 x 2</span>
<span class="co">#&gt;    .win_id total_coverage</span>
<span class="co">#&gt;      &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1      0.5370370</span>
<span class="co">#&gt;  2       2      3.2555556</span>
<span class="co">#&gt;  3       3      0.4534884</span>
<span class="co">#&gt;  4       4      1.4900000</span>
<span class="co">#&gt;  5       5      1.0415094</span>
<span class="co">#&gt;  6       6      4.3157895</span>
<span class="co">#&gt;  7       7      0.8145161</span>
<span class="co">#&gt;  8       8      4.1476923</span>
<span class="co">#&gt;  9       9      8.9347222</span>
<span class="co">#&gt; 10      10      1.8328571</span>
<span class="co">#&gt; # ... with 191 more rows</span></code></pre></div>
</div>
<div id="combining-many-tibbles" class="section level2">
<h2 class="hasAnchor">
<a href="#combining-many-tibbles" class="anchor"></a>Combining many tibbles</h2>
<p><code>bind_rows()</code> can bind any number of tibbles if the tibbles are supplied as a <code>list()</code>. This is very useful to combining many dataframes that are in a list into a format for plotting.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">all_res &lt;-<span class="st"> </span><span class="kw">bind_rows</span>(res, <span class="dt">.id =</span> <span class="st">"expt"</span>)
all_res
<span class="co">#&gt; # A tibble: 603 x 3</span>
<span class="co">#&gt;     expt .win_id total_coverage</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt;  1     1       1      0.8479167</span>
<span class="co">#&gt;  2     1       2      4.7904762</span>
<span class="co">#&gt;  3     1       3      0.9065217</span>
<span class="co">#&gt;  4     1       4      4.7795918</span>
<span class="co">#&gt;  5     1       5      4.1403846</span>
<span class="co">#&gt;  6     1       6      2.3000000</span>
<span class="co">#&gt;  7     1       7      6.1677419</span>
<span class="co">#&gt;  8     1       8      2.9953846</span>
<span class="co">#&gt;  9     1       9      7.5516667</span>
<span class="co">#&gt; 10     1      10      0.9807018</span>
<span class="co">#&gt; # ... with 593 more rows</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(all_res, <span class="kw">aes</span>(.win_id, total_coverage)) +
<span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">colour =</span> expt))</code></pre></div>
<p><img src="img/class-2-plot_random-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div id="exercises-1" class="section level2">
<h2 class="hasAnchor">
<a href="#exercises-1" class="anchor"></a>Exercises</h2>
<p>The exercises for today’s class are presented <a href="https://rnabioco.github.io/eda/articles/exercises-4.html">here</a>, Here is a <a href="https://raw.githubusercontent.com/rnabioco/eda/master/vignettes/exercises-4.Rmd">link</a> to the raw text of these exercises</p>
</div>
<div id="quiz" class="section level2">
<h2 class="hasAnchor">
<a href="#quiz" class="anchor"></a>Quiz</h2>
<p>Create an RMarkdown document and answer the questions from the Exercises section. Write your answers in the text, and print the final tibbles that your code produced that gave you the answer. Submit your final document as “Quiz 4” by <strong>Tuesday at 10 PM</strong>.</p>
<p><strong>Your submitted document must knit to HTML without errors</strong>. I.e., when you click the “Knit” button, the document should build and display and HTML page.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#genomic-interval-data-pt-2">Genomic Interval Data pt.2</a><ul class="nav nav-pills nav-stacked">
<li><a href="#generating-random-intervals-using-bed_shuffle-and-bed_random">Generating random intervals using <code><a href="http://www-rdocumentation-org/packages/valr/topics/bed_shuffle">bed_shuffle()</a></code> and <code><a href="http://www-rdocumentation-org/packages/valr/topics/bed_random">bed_random()</a></code></a></li>
      <li><a href="#exercises">Exercises</a></li>
      <li><a href="#how-to-generate-a-metagene-plot-with-multiple-samples">How to generate a metagene plot with multiple samples?</a></li>
      <li><a href="#bind_rows-binds-rows-of-multiple-tibbles-together">bind_rows(), binds rows of multiple tibbles together</a></li>
      <li><a href="#generating-a-metagene-with-multiple-samples">Generating a metagene with multiple samples</a></li>
      <li><a href="#writing-a-function-to-produce-a-metagene">Writing a function to produce a metagene</a></li>
      <li><a href="#writing-functions-reduces-repeated-code">Writing functions reduces repeated code</a></li>
      <li><a href="#introduction-to-functions-in-r">Introduction to functions in R</a></li>
      <li><a href="#functions-can-have-default-arguments">Functions can have default arguments</a></li>
      <li><a href="#computing-the-metagene-profile-using-a-function">Computing the metagene profile using a function</a></li>
      <li><a href="#using-lists-to-organize-many-objects">Using lists to organize many objects</a></li>
      <li><a href="#lists-make-it-easy-to-store-and-organize-many-objects">Lists make it easy to store and organize many objects</a></li>
      <li><a href="#the-map-function">The map function</a></li>
      <li><a href="#combining-many-tibbles">Combining many tibbles</a></li>
      <li><a href="#exercises-1">Exercises</a></li>
      <li><a href="#quiz">Quiz</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Austin Gillen, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
